File: Serenity.Script.UI/FilterPanel/FilterLine.cs
using Serenity.Data;
using System;
using System.Runtime.CompilerServices;

namespace Serenity
{
    [Imported, Serializable]
    public class FilterLine
    {
        public string Field { get; set; }
        public string Operator { get; set; }
        public bool IsOr { get; set; }
        public bool LeftParen { get; set; }
        public bool RightParen { get; set; }
        public string ValidationError { get; set; }
        public BaseCriteria Criteria { get; set; }
        public string DisplayText { get; set; }
        public object State { get; set; }
    }
}

File: packages/corelib/src/ui/filtering/filterline.ts
export { };

/**
 * A single line in a filter panel, describing one filter condition.
 */
export interface FilterLine {
    /** Field name being filtered. */
    field?: string;
    /** Operator key. */
    operator?: string;
    /** Whether this line is OR-combined with the previous line. */
    isOr?: boolean;
    /** Whether this line opens a parenthesis group. */
    leftParen?: boolean;
    /** Whether this line closes a parenthesis group. */
    rightParen?: boolean;
    /** Validation error message, if any. */
    validationError?: string;
    /** The criteria expression for this line. */
    criteria?: any[];
    /** Display text for this line. */
    displayText?: string;
    /** Persisted editor state. */
    state?: any;
}